JointMeasurements/Tests.qs

```diff
@@ -20,42 +20,71 @@ namespace Quantum.Kata.JointMeasurements {
     
     // "Framework" operation for testing multi-qubit tasks for distinguishing states of an array of qubits
     // with Int return
-    operation DistinguishStates_MultiQubit (nQubit : Int, nState : Int, statePrep : ((Qubit[], Int, Double) => Unit is Adj), testImpl : (Qubit[] => Int), preserveState : Bool) : Unit {
+    operation DistinguishStates_MultiQubit (nQubits : Int,
+                                            nStates : Int,
+                                            statePrep : ((Qubit[], Int, Double) => Unit is Adj),
+                                            testImpl : (Qubit[] => Int),
+                                            measurementsPerRun : Int,
+                                            stateNames : String[]) : Unit {
         let nTotal = 100;
-        mutable nOk = 0;
-        
-        using (qs = Qubit[nQubit]) {
+        // misclassifications will store the number of times state i has been classified as state j (dimension nStates^2)
+        mutable misclassifications = new Int[nStates * nStates];
+        // unknownClassifications will store the number of times state i has been classified as some invalid state (index < 0 or >= nStates)
+        mutable unknownClassifications = new Int[nStates];
+                
+        using (qs = Qubit[nQubits]) {
             for (i in 1 .. nTotal) {
                 // get a random integer to define the state of the qubits
                 let state = DrawRandomInt(0, nState - 1);
-                
                 // get a random rotation angle to define the exact state of the qubits
                 let alpha = DrawRandomDouble(0.0, 1.0) * PI();
                 
                 // do state prep: convert |0...0⟩ to outcome with return equal to state
                 statePrep(qs, state, alpha);
-                
-                // get the solution's answer and verify that it's a match
+
+                if (measurementsPerRun > 0) {
+                    ResetOracleCallsCount();
+                }
+                // get the solution's answer and verify that it's a match, if not, increase the exact mismatch count
                 let ans = testImpl(qs);
-                if (ans == state) {
-                    set nOk += 1;
+                if ((ans >= 0) and (ans < nStates)) {
+                    // classification result is a valid state index - check if is it correct
+                    if (ans != state) {
+                        set misclassifications w/= ((state * nStates) + ans) <- (misclassifications[(state * nStates) + ans] + 1);
+                    }
                 }
-                
-                if (preserveState) {
```

Comment:
Please keep the logic of `preserveState` variable - you'll notice that the Measurements kata never requires the learner to preserve the input state while learning which state it was, but JointMeasurements kata asks to do this in multiple tasks